rust/lance/src/dataset/write/merge_insert/exec/indexed_lookup.rs

```diff
@@ -0,0 +1,470 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! Index-based merge join execution node.
+//!
+//! This node uses a btree index to efficiently perform merge-insert lookups.
+//! It has two modes of operation:
+//!
+//! ## Mode 1: Lookup Only (join_type = None)
+//!
+//! Used for hybrid execution where the join happens externally.
+//! Input: key projection (just the key column from source)
+//! Output: matching target rows [projection_columns, _rowid, _rowaddr]
+//!
+//! ## Mode 2: Lookup with Join (join_type = Some)
+//!
+//! Used for fully indexed execution where the join is internal.
+//! Input: full source data
+//! Output: joined rows [source_cols, target_cols with _rowid, _rowaddr]
+//!
+//! The internal DAG structure (ReplayExec used twice) is hidden from DataFusion
+//! by only exposing the original input as a child. This prevents DataFusion's
+//! tree transformations from breaking the shared ReplayExec references.
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow_schema::{Field, Schema, SchemaRef};
+use datafusion::common::NullEquality;
+use datafusion::error::Result as DFResult;
+use datafusion::execution::{SendableRecordBatchStream, TaskContext};
+use datafusion::physical_plan::execution_plan::{Boundedness, EmissionType};
+use datafusion::physical_plan::joins::{HashJoinExec, PartitionMode};
+use datafusion::physical_plan::metrics::{ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::physical_plan::projection::ProjectionExec;
+use datafusion::physical_plan::{DisplayAs, ExecutionPlan, Partitioning, PlanProperties};
+use datafusion_physical_expr::expressions::Column;
+use datafusion_physical_expr::EquivalenceProperties;
+use lance_core::utils::futures::Capacity;
+use lance_core::{ROW_ADDR_FIELD, ROW_ID_FIELD};
+use lance_table::format::IndexMetadata;
+
+use crate::io::exec::scalar_index::MapIndexExec;
+use crate::io::exec::utils::ReplayExec;
+use crate::io::exec::{project, AddRowAddrExec, TakeExec};
+use crate::Dataset;
+
+/// Index-based lookup execution node.
+///
+/// This node uses a btree index to efficiently lookup target rows.
+///
+/// ## Modes
+///
+/// - **Lookup only** (`join_type = None`): Takes key projection as input,
+///   outputs matching target rows.
+/// - **Lookup with join** (`join_type = Some`): Takes source as input,
+///   outputs joined rows with source and target columns.
+///
+/// ## Output Schema
+///
+/// Lookup only mode:
+/// - `projection_columns` from target
+/// - `_rowid`
+/// - `_rowaddr` (if with_row_addr)
+///
+/// Join mode:
+/// - Source columns
+/// - Target columns (including `_rowid`, `_rowaddr`)
+#[derive(Debug)]
+pub struct IndexedLookupExec {
```

Comment:
I ended up wrapping this pipeline of source -> project -> MapIndexExec -> AddRowAddrExec -> TakeExec -> project as this execution node, since otherwise DF optimizer keeps adding additional repartitioning steps in between. Not sure if there is any better way.

Reply:
> otherwise DF optimizer keeps adding additional repartitioning steps in between

I think the DataFusion optimizer does this based on [ExecutionPlan::benefits_from_input_partitioning](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.benefits_from_input_partitioning) and [ExecutionPlan::required_input_distribution](https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#method.required_input_distribution). It compares those settings with the children nodes input partitioning to decide whether to add repartitioning. So also worth making sure you are stating the partitioning correctly.